File: ZendriveSDKDemo/UserDefaultsManager.m
//
//  UserDefaultsManager.m
//  ZendriveSDKDemo
//
//  Created by Yogesh on 4/11/15.
//  Copyright (c) 2015 Zendrive. All rights reserved.
//

#import "UserDefaultsManager.h"
#import "User.h"
#import "Trip.h"

#define UserDefaults [NSUserDefaults standardUserDefaults]

static NSString *kLoggedInUserKey = @"loggedInUser";
static NSString *kDriveDetectionModeKey = @"driveDetectionMode";
static NSString *kServiceTierKey = @"serviceTier";
static NSString *kTripsUserDefaultsKey = @"tripsArray";
static UserDefaultsManager *_sharedInstance;

@implementation UserDefaultsManager

+ (instancetype)sharedInstance {
    if (_sharedInstance == nil) {
        @synchronized (self) {
            if (_sharedInstance == nil) {
                _sharedInstance = [[self alloc] init];
            }
        }
    }
    return _sharedInstance;
}

//------------------------------------------------------------------------------
#pragma mark - Logged In User
//------------------------------------------------------------------------------
- (void)setLoggedInUser:(User *)user {
    [UserDefaults setObject:user.toDictionary forKey:kLoggedInUserKey];
    [UserDefaults synchronize];
}

- (User *)loggedInUser {
    NSDictionary *userDictionary = [UserDefaults objectForKey:kLoggedInUserKey];
    if (userDictionary) {
        return [[User alloc] initWithDictionary:userDictionary];
    }
    return nil;
}

//------------------------------------------------------------------------------
#pragma mark - Drive detection mode
//------------------------------------------------------------------------------
- (void)setDriveDetectionMode:(ZendriveDriveDetectionMode)driveDetectionMode {
    [UserDefaults setObject:@(driveDetectionMode) forKey:kDriveDetectionModeKey];
    [UserDefaults synchronize];
}

- (ZendriveDriveDetectionMode)driveDetectionMode {
    NSNumber *driveDetectionModeNsNum = [UserDefaults objectForKey:kDriveDetectionModeKey];
    if (driveDetectionModeNsNum) {
        return driveDetectionModeNsNum.intValue;
    }
    return ZendriveDriveDetectionModeAutoON;
}

//------------------------------------------------------------------------------
#pragma mark - Service Tier
//------------------------------------------------------------------------------
- (void)setServiceTier:(int)serviceTier {
    [UserDefaults setObject:@(serviceTier) forKey:kServiceTierKey];
    [UserDefaults synchronize];
}

- (int)serviceTier {
    NSNumber *serviceTierNsNum = [UserDefaults objectForKey:kServiceTierKey];
    if (serviceTierNsNum) {
        return serviceTierNsNum.intValue;
    }
    return ZendriveServiceLevelDefault;
}

//------------------------------------------------------------------------------
#pragma mark - Saving and retrieving trips
//------------------------------------------------------------------------------
- (void)saveTrip:(Trip *)trip {
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    NSDictionary *tripDictionary = [trip toDictionary];
    NSArray *tripsArray = [userDefaults objectForKey:kTripsUserDefaultsKey];
    NSArray *newTripsArray;
    if (tripsArray == nil) {
        newTripsArray = @[tripDictionary];
    }
    else {
        newTripsArray = [tripsArray arrayByAddingObject:tripDictionary];
    }

    [userDefaults setObject:newTripsArray forKey:kTripsUserDefaultsKey];
    [userDefaults synchronize];
}

- (void)updateTrip:(Trip *)trip {
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    NSDictionary *tripDictionary = [trip toDictionary];
    NSMutableArray *tripsArray = [[userDefaults objectForKey:kTripsUserDefaultsKey] mutableCopy];
    if (tripsArray != nil) {
        int i = -1;
        for (i = 0; i < tripsArray.count; i++) {
            NSDictionary *thisTripDict = tripsArray[i];
            Trip *thisTrip = [[Trip alloc] initWithDictionary:thisTripDict];
            if ([thisTrip.startDate compare:trip.startDate] == NSOrderedSame) {
                break;
            }
        }

        if (i >= 0 && i < tripsArray.count) {
            [tripsArray replaceObjectAtIndex:i withObject:tripDictionary];
        }
    }

    [userDefaults setObject:tripsArray forKey:kTripsUserDefaultsKey];
    [userDefaults synchronize];
}

- (NSMutableArray *)fetchAllTrips {
    NSMutableArray *tripsArray = [[NSMutableArray alloc] init];

    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    NSArray *tripDictionariesArray = [userDefaults objectForKey:kTripsUserDefaultsKey];
    for (NSDictionary *tripDictioanry in tripDictionariesArray) {
        Trip *trip = [[Trip alloc] initWithDictionary:tripDictioanry];
        [tripsArray insertObject:trip atIndex:0];
    }

    return tripsArray;
}
@end

File: swift/ZendriveSDKDemoSwift/UserDefaultsManager.swift
//
//  UserDefaultsManager.swift
//  ZendriveSDKDemoSwift
//
//  Created by amrit on 19/07/19.
//  Copyright © 2019 zendrive. All rights reserved.
//

import Foundation
import ZendriveSDK

private var _sharedInstance: UserDefaultsManager? = nil

final class UserDefaultsManager {
    private let loggedInUserKey = "loggedInUser"
    private let driveDetectionModeKey = "driveDetectionMode"
    private let serviceTierKey = "serviceTier"
    private let tripsUserDefaultsKey = "tripsArray"

    class func sharedInstance() -> UserDefaultsManager {
        if _sharedInstance == nil {
            let lockQueue = DispatchQueue(label: "self")
            lockQueue.sync {
                if _sharedInstance == nil {
                    _sharedInstance = UserDefaultsManager()
                }
            }
        }
        return _sharedInstance!
    }

    func loggedInUser() -> [AnyHashable: Any]? {
        let userDictionary = UserDefaults.standard.object(forKey: loggedInUserKey) as? [AnyHashable : Any]
        return userDictionary
    }

    func setLoggedIn(_ user: User?) {
        UserDefaults.standard.set(user?.toDictionary(), forKey: loggedInUserKey)
    }

    func driveDetectionMode() -> ZendriveDriveDetectionMode {
        let driveDetectionModeNsNum: Int = UserDefaults.standard.integer(forKey: driveDetectionModeKey)
        return ZendriveDriveDetectionMode(rawValue: Int32(driveDetectionModeNsNum))!
    }

    func setDriveDetectionMode(_ driveDetectionMode: ZendriveDriveDetectionMode) {
        UserDefaults.standard.set(driveDetectionMode.rawValue, forKey: driveDetectionModeKey)
        UserDefaults.standard.synchronize()
    }

    func serviceTier() -> Int {

        if let serviceTierNsNum = UserDefaults.standard.object(forKey: serviceTierKey) as? Int {
            return serviceTierNsNum
        }

        return 0
    }

    func setServiceTier(_ serviceTier: Int) {
        UserDefaults.standard.set(serviceTier, forKey: serviceTierKey)
        UserDefaults.standard.synchronize()
    }

    //------------------------------------------------------------------------------
    // MARK: - Trip load/save
    //------------------------------------------------------------------------------
    func save(_ trip: Trip) {
        let userDefaults = UserDefaults.standard
        var newTripDictArray: [[AnyHashable: Any]] = []

        if let tripDictArray = userDefaults.object(forKey: tripsUserDefaultsKey) as? [[AnyHashable: Any]] {
            let array = [trip.toDictionary()] + tripDictArray
            newTripDictArray = array
        } else {
            newTripDictArray = [trip.toDictionary()]
        }
        userDefaults.set(newTripDictArray, forKey: tripsUserDefaultsKey)
        userDefaults.synchronize()
    }

    func update(_ trip: Trip?) {
        let userDefaults = UserDefaults.standard
        let tripDictionary = trip?.toDictionary()
        var tripDictArray = userDefaults.object(forKey: tripsUserDefaultsKey) as? [[AnyHashable: Any]]
        if tripDictArray != nil {
            var index = -1
            for i in 0..<(tripDictArray?.count ?? 0) {
                let thisTripDict = tripDictArray?[i]
                let thisTrip = Trip(dictionary: thisTripDict)
                if thisTrip.startDate == trip?.startDate {
                    index = i
                    break
                }
            }

            if index >= 0 && index < (tripDictArray?.count ?? 0) {
                if let tripDictionary = tripDictionary {
                    tripDictArray?[index] = tripDictionary
                }
            }
        }

        userDefaults.set(tripDictArray, forKey: tripsUserDefaultsKey)
        userDefaults.synchronize()
    }

    func fetchAllTrips() -> [Trip]? {
        var tripsArray: [Trip] = []
        let userDefaults = UserDefaults.standard
        let tripDictArray = userDefaults.object(forKey: tripsUserDefaultsKey) as? [[AnyHashable: Any]]

        if let _ = tripDictArray {
            for i in 0..<(tripDictArray?.count ?? 0) {
                tripsArray.append(Trip(dictionary: tripDictArray![i]))
            }
        }

        return tripsArray
    }

}
